solana/gatekeeper-lib/src/service/GatekeeperService.ts

```diff
@@ -152,14 +205,29 @@ export class GatekeeperService {
       )
     );
 
-    await send(
-      this.connection,
-      transaction,
-      this.retries,
-      confirmOptions,
-      this.payer,
-      this.gatekeeperAuthority
-    );
+    try {
+      await send(
+        this.connection,
+        transaction,
+        confirmOptions,
+        this.payer,
+        this.gatekeeperAuthority
+      );
+    } catch (err) {
+      // The tx may have succeeded but the blockchain call timed out. Check on-chain.
+      const onChainToken: GatewayToken | null = await getGatewayToken(
+        this.connection,
+        gatewayTokenKey
+      );
+      if (onChainToken?.state === State.REVOKED) {
```

Comment:
I think this logic could be abstracted into a function `checkStateOrThrow` or something like that where the token and expected state are passed

Reply:
or even passing a predicate instead of desired state to handle the other cases below

Reply:
Done